File: C++/src/print/print.cpp
#include <bits/stdc++.h>
#include "../inventario/inventario.h"
#include "../estruturas/estruturas.h"
#include "../fases/fases.h"
#include "../batalha/batalha.h"

#include "print.h"

using namespace std;

int getChoice(Personagem &p){
    printf("\n\n\n");
    StatusHeroi(p);
    printf("[1] -> Fases\n");
    printf("[2] -> Loja\n");
    printf("[3] -> Bolsa\n");
    printf("[4] -> Baú\n");
    printf("[5] -> Créditos\n");
    printf("[6] -> Sair\n");
	int opcao;
    printf("\n\nDigite sua opcao: ");
    scanf("%d",&opcao);
    return opcao;
}

int interfaceFases() {
    printf("[1] -> Manguezal\n");
    printf("[2] -> Casa\n");
    printf("[3] -> Jogoses Voraz\n");
    printf("[4] -> Piloto\n");
    printf("[5] -> Barquinho\n");
    printf("[6] -> BOSS\n");
    printf("[7] -> sair");
    printf("\n\nEscolha sabiamente a fase desejada... ");

    int opcao;
    scanf("%d", &opcao);
    return opcao;
}


int interfaceMapa(Fase &f) {
    printf("Você está no mapa: \n");

    cout << endl << f.nome << endl << endl;
    cout <<  "    +-> " << f.descricao << endl;
    printf("\n\n\n\n\n");

    printf("[1] -> Entrar em uma batalha\n");
    printf("[2] -> Vasculhar sua bolsa\n");
    printf("[3] -> Pausa pro café\n");

    printf("\nQual a sua escolha?  ");

    int opcao;
    scanf("%d", &opcao);
    printf("\n");
    return opcao;
}

int escolhasDaBolsa(){
    printf("[1] -> Visualizar poções\n");
    printf("[2] -> Usar uma poção\n");
    printf("[3] -> Jogar uma poção fora\n");
    printf("[4] -> Voltar a batalhar\n");
    printf("\nO que você quer fazer?... ");
    int opcao;
    scanf("%d", &opcao);
    return opcao;
}

void StatusHeroi(Personagem &heroi) {
    cout << heroi.nome;
    for(int i = 0; i < (30 - heroi.nome.size()); i++)
        cout << " ";
    cout << heroi.vidaAtual << "/" << heroi.vidaTotal << endl << endl;
}

void HpCombate(Personagem &p, GrupoDeInimigos gp) {
    cout << endl;
    StatusHeroi(p);
    for (int i = 0; i < gp.quantidade; i++) {
        cout << gp.gangue[i].nome;
        for(int j = 0; j < (30 - gp.gangue[i].nome.size()); j++)
            cout << " ";
        cout  << gp.gangue[i].vidaAtual << "/" << gp.gangue[i].vidaTotal << endl;
    }
}

int opcoesAtaque() {
    int opcao;
    do{
        printf("\n\n\n\n\n");
        printf("Como você quer atacar?\n\n");
        printf("[1] -> Ataque Forte\n");
        printf("[2] -> Ataque Fraco\n");
        printf("\nDigite sua opção: ");

        scanf("%d", &opcao);

        switch (opcao) {
            case 1: {
                printf("\nVocê selecionou ataque FORTE\n");
                break;
            } case 2: {
                printf("\nVocê selecionou ataque FRACO\n");
                break;
            }
        }
    }while(opcao != 1 && opcao != 2);

    return opcao;
}

int ataqueInimigo(GrupoDeInimigos gp){
    int opcao;
    do{
        printf("\n\n\n\n");

        for(int i = 1; i <= gp.quantidade; i++){
            cout << "atacar: " << gp.gangue[i-1].nome << " [" << i << "]" << " ";

            if (gp.gangue[i - 1].vidaAtual <= 0) {
                cout << "MORTO\n\n";
            } else {
                cout << gp.gangue[i - 1].vidaAtual << "/" << gp.gangue[i - 1].vidaTotal << "\n\n";
            }
        }
        printf("Digite quem você quer atacar: ");
        scanf("%d",&opcao);

    }while(opcao < 1 || opcao > gp.quantidade);
}

void wonBattle(Personagem &p, GrupoDeInimigos gp) {
    cout << "Parabéns " << p.nome << "." << endl;
    printf("Você ganhou %d moedas.\n\n", gp.dinheiroLoot);
}

void lostBattle(GrupoDeInimigos gp) {
    printf("Oxe doido tu perdeu feião visse ;(\n");
    cout << "Você perdeu " << gp.dinheiroLoot * 0.1 << " moedas.\n\n";
}

void digite(){
    setbuf(stdin,NULL);
    printf("\nDigite enter para continuar...");
    getchar();
}

void enter() {
    setbuf(stdin,NULL);
    getchar();
}

void inicioDoJogo() {
    printf("Tecle enter para prosseguir\n\n");
    enter();

    printf("Você acorda em um local que você nunca viu antes,");
    enter();
    printf("somente com as roupas do seu corpo.");
    enter();
    printf("Você não sabe onde está...");
    enter();
    printf("Tudo o que você sabe é o que você leu no bilhete que estava na sua mão quando acordou...");
    enter();
    printf("Para escapar desse mundo, você deve derrotar o temível *insira o nome aqui*.\n");
    enter();

    printf("Você lembra do seu nome?...  ");
}

void divisorias() {
    printf("////////////////////////////////////////////////////////////////////////////////\n");
}

int escolhasDoBau(){
  printf("[1] -> Troque seu equipamento\n");
  printf("[2] -> Excluir um item do Baú\n");
  printf("[3] -> Visualizar equipamento\n");
  printf("[4] -> Vasculhar sua bolsa\n");
  printf("[5] -> Organizar sua bolsa\n");
  printf("[6] -> Voltar ao menu inicial\n");
  int opcao;
  printf("\nO que você vai querer hoje? ");
  scanf("%d", &opcao);
  return opcao;
}

void printInimigos(GrupoDeInimigos &gp){
    system("clear");
    printf("Conheça seus inimigos:\n\n");
    for(int i = 0; i < gp.quantidade; i++){
        Inimigo ini = gp.gangue[i];
        cout << ini.nome << " " << ini.vidaAtual << "/" << ini.vidaTotal << endl;
        cout << "    +-> " << ini.descricao << "\n\n";
    }
    digite();
}

void estrelinhas(){
    printf("*****************************");
}


int printLoja(Personagem &p){
    system("clear");
    setbuf(stdin,NULL);
    cout << "Você tem : " << p.dinheiro << " de dinheiro" << endl;
    int opcao;
    do{
        printf("#       LOJA      #\n\n");
        printf("[1] -> Comprar uma arma\n");
        printf("[2] -> Comprar uma armadura\n");
        printf("[3] -> Comprar uma poção\n");
        printf("[4] -> sair\n\n");
        printf("Digite sua opção: ");
        scanf("%d",&opcao);
    }while(opcao < 0 || opcao > 4);
    return opcao;
}

int listaArma(Loja &l){
    int opcao;
    setbuf(stdin,NULL);
    do{
        for(int i = 1; i <= l.quantArmas; i++){
            Arma a = l.armas[i-1];
            cout << "["<<  i << "] -> nome: " << a.nome << endl;
            cout << "       descrição: " << a.descricao << endl;
            cout << "       (dano: " << a.dano << "/ força: " << a.forca << "/ agilidade: " << a.agilidade << ")\n";
            cout << "       preço: " << a.preco << endl;
        }
        printf("digite sua opcao: ");
        scanf("%d",&opcao);
    }while(opcao < 1 || opcao > l.quantArmas);
    return opcao;
}


int listaArmadura(Loja &l){
    int opcao;
    setbuf(stdin,NULL);
    do{
        for(int i = 1; i <= l.quantArmaduras; i++){
            Armadura a = l.armaduras[i-1];
            cout << "["<<  i << "] -> nome: " << a.nome << endl;
            cout << "       descrição: " << a.descricao << endl;
            cout << "       ( armadura: " << a.armadura << "/ força: " << a.forca << "/ agilidade: " << a.agilidade << "/ vida: " << a.vida <<")\n";
            cout << "       preço: " << a.preco << endl;
        }
        printf("digite sua opcao: ");
        scanf("%d",&opcao);
    }while(opcao < 1 || opcao > l.quantArmaduras);
    return opcao;
}

int listaPocoes(Loja &l){
    int opcao;
    setbuf(stdin,NULL);
    do{
        for(int i = 1; i <= l.quantPocoes; i++){
            Pocao a = l.pocoes[i-1];
            cout << "["<<  i << "] -> nome: " << a.nome << endl;
            cout << "       descrição: " << a.descricao << endl;
            cout << "       + Vida-> (" << a.vida <<")\n";
            cout << "       preço: " << a.preco << endl;

        }
        printf("Digite sua opçao");
        scanf("%d",&opcao);
    }while(opcao < 1 || opcao > l.quantPocoes);
    return opcao;
}

void listaPocoes(Bau &bau) {
	if (bau.indPocoes < 0) {
		printf("Você não tem poções! Vá tomar água enquanto isso.\n\n");
	} else {
		for (int i = 1; i <= bau.indPocoes + 1; i++) {
			Pocao p = bau.pocoes[i - 1];
			cout << "[" << i << "] -> nome: " << p.nome << endl;
			cout << "       descrição: " << p.descricao << endl;
			cout << "       (";
			verificaStatus(p.vida);
			cout << ")\n";
			cout << "       preço: " << p.preco << endl;
		}
	}
}

void verificaStatus(int status) {
	if (status >= 0)
		cout << "+" << status;
	else
		cout << status;
}

int opcoesBolsaBau() {
	int opcao;
	printf("Como você deseja organizar a sua bolsa?\n\n");
	printf("[1] -> Mover poção do Baú para a Bolsa\n");
	printf("[2] -> Mover poção da Bolsa para o Baú\n");
	printf("[3] -> Cancelar\n");

    printf("O que você quer fazer? ");
	scanf("%d", &opcao);
	return opcao;
}

int escolhaUmaPocao(int range) {
    int opcao = -1;

    do {
        printf("[0] -> Cancelar\n\n");
        printf("Digite o número da poção escolhida [1, %d] ", range + 1);
        scanf("%d", &opcao);

        if (opcao < 0 || opcao > range + 1) {
            printf("Esse número não fez sentido. Tente de novo\n");
        }
    } while (opcao < 0 || opcao > range + 1);

    return opcao - 1; // retorna como índice pronto para ser usado no array
}

int removePocao(int range) {
    int opcao = -1;

    do {
        printf("[0] -> Cancelar\n\n");
        printf("Digite o índice da poção que deseja remover: ");
        scanf("%d", &opcao);

        if (opcao < 0 || opcao > range + 1) {
            printf("Esse número não fez sentido. Tente de novo\n");
        }
    } while (opcao < 0 || opcao > range + 1);

    return opcao - 1;
}

int tomaPocao(Personagem &p) {
    int opcao = -1;

    do {
        printf("\n[0] -> Cancelar\n\n");
        listBag(p);
        printf("\nQual poção você quer usar?  \n");
        scanf("%d", &opcao);

        if (opcao > p.bolsa.quantidade || opcao < 0) {
            printf("Índice inválido. Tente de novo..");
        }
    } while (opcao > p.bolsa.quantidade || opcao < 0);

    return opcao - 1;
}

int excluirItem() {
    int opcao = -1;

    do {
        printf("[1] -> Excluir uma Armadura\n");
        printf("[2] -> Excluir uma Arma\n");
        printf("[3] -> Excluir uma Poção\n");
        printf("[4] -> Sair\n");
        printf("\nQual a sua opção? ");
        scanf("%d", &opcao);

        if (opcao < 1 || opcao > 4)
            printf("Opção inválida. Tente de novo\n");
    } while (opcao < 1 || opcao > 4);

    return opcao;
}

int trocarEquipamento() {
    int opcao = -1;

    do {
        printf("[1] -> Trocar uma Arma\n");
        printf("[2] -> Trocar uma Armadura\n");
        printf("[3] -> Sair\n");

        printf("\nQual a sua opção? ");
        scanf("%d", &opcao);

        if (opcao < 1 || opcao > 3)
            printf("Opção inválida. Tente de novo\n");
    } while (opcao < 1 || opcao > 3);

    return opcao;
}

File: Haskell/Print.hs
module Print where
import Estruturas
import Entidades
import System.IO.Unsafe
import System.Process
import System.IO


printaVida :: Personagem -> IO ()
printaVida p =  do
                putStrLn ((personagemNome p ) ++ "                   " ++ (show (personagemVidaAtual p)) ++ "/" ++ (show (personagemVidaMax p)) ++ "\n")

printaFases :: Personagem -> IO Int
printaFases per =   do
                    system "clear"
                    putStrLn "              #   MAPAS   #\n\n"
                    putStrLn "Com grandes escolhas vem grande responsabilidades.\nEscolha sabiamente.\n"
                    putStrLn "[1] -> Manguezal"
                    putStrLn "[2] -> Casa"
                    putStrLn "[3] -> Jogoses Voraz"
                    putStrLn "[4] -> Piloto"
                    putStrLn "[5] -> Área 51"
                    putStrLn "[6] -> BOSS"
                    putStrLn "[7] -> sair"
                    putStr   "\nEscolha sabiamente a fase desejada... "
                    a <- readLn :: IO Int
                    return a

fn :: Personagem -> IO Personagem
fn p = do
    if (personagemNome p == "") then
        return p
    else 
        return p

printLobby :: Personagem -> IO Int
printLobby p =  do
                printaVida p  
                putStrLn "\n              #   LOBBY   #\n\n\n"
                putStrLn "[1] -> Fases"
                putStrLn "[2] -> Loja"
                putStrLn "[3] -> Bolsa"
                putStrLn "[4] -> Baú"
                putStrLn "[5] -> Créditos"
                putStrLn "[6] -> Sair"
                putStr "\n\nDigite sua opção: "
                a <- readLn :: IO Int
                return a

printLoja :: Personagem -> IO Int
printLoja p =  do
                    putStrLn "              #       LOJA      #\n\n"
                    putStrLn "[1] -> Comprar uma arma"
                    putStrLn "[2] -> Comprar uma armadura"
                    putStrLn "[3] -> Comprar uma poção"
                    putStrLn "[4] -> sair\n\n"
                    putStr "Digite sua escolha: "
                    a <- readLn :: IO Int
                    return a



boasVindas :: IO()
boasVindas =    do
                system "clear"
                putStrLn "Tecle enter para prosseguir.\n"
                enter

                putStrLn "Você acorda em um local que você nunca viu antes,"
                enter
                putStrLn "somente com as roupas do seu corpo."
                enter
                putStrLn "Você não sabe onde está..."
                enter
                putStrLn "Tudo o que você sabe é o que você \nleu no bilhete que estava na sua mão quando acordou..."
                enter
                putStrLn "Para escapar desse mundo, você deve derrotar a temível LIGHT THEME IDE.\n"
                enter
                hSetEcho stdin True     -- volta a mostrar no terminal o que o user esta escrevendo
                putStr "Você lembra do seu nome?... "


printMapa :: IO Int
printMapa = do
    putStr "\n"
    putStrLn "[1] -> Entrar em uma batalha"
    putStrLn "[2] -> Vasculhar sua bolsa"
    putStrLn "[3] -> Pausa para o café"
    putStr "\nQual a sua escolha? "

    a <- readLn :: IO Int
    return a

descrMapa :: Fase -> IO()
descrMapa mapa = do
    system "clear"
    putStrLn "Você está no mapa:\n"
    putStrLn $ (faseNome mapa) ++ "\n"
    putStrLn $ "    +-> " ++ (faseDescricao mapa)
    putStr "\n\n\n\n"


estrelinhas :: IO()
estrelinhas = do
    putStr "****************************************\n"

divisorias :: IO()
divisorias = do
    putStr "////////////////////////////////////////////////////////////////////////////////\n"

fazNadaPer :: Personagem -> Personagem
fazNadaPer p = p

atkCritico :: IO()
atkCritico = do
    putStrLn "Ataque Crítico!  "

printHeroiAtaca :: Personagem -> Int -> IO()
printHeroiAtaca p dano = do
    putStrLn $ "\n" ++ (personagemNome p) ++ " deu " ++ (show dano) ++ " de dano."

fazNada :: Int -> IO ()
fazNada a = do
    if(a == -4675859550) then
        print a
    else
        putStr ""

printInimigoAtaca :: Inimigo -> Int -> IO()
printInimigoAtaca i dano = do
    putStrLn ((inimigoNome i) ++ " deu " ++ (show (inimigoDano i)) ++ " de dano.")

printHeroiTomaDano :: Personagem -> Int -> IO()
printHeroiTomaDano p dano = do
    putStrLn ((personagemNome p) ++ " recebeu " ++ (show dano) ++ " de dano.")

printInimigoTomaDano :: Inimigo -> Int -> IO()
printInimigoTomaDano i dano = do
    if ((inimigoVidaAtual i) <= 0) then
        putStrLn "Você tá chutando cachorro morto..."
    else
        putStrLn ( (inimigoNome i) ++ " recebeu " ++ (show dano) ++ " de dano." )

lostBattle :: Int -> IO ()
lostBattle per = do
    putStr "\n\n"
    estrelinhas
    putStrLn "Oxe doido tu perdeu feião visse ;(\n"
    putStrLn ("Você perdeu " ++ (show per) ++ " moedas.")
    estrelinhas
    putStrLn "\n"

wonBattle :: Personagem -> Int -> IO ()
wonBattle per num = do
    putStr "\n\n"
    estrelinhas
    putStrLn ("Parabéns " ++ (personagemNome per) ++ ".\n")
    putStrLn ("Você ganhou " ++ (show num) ++ " moedas.")
    estrelinhas
    putStrLn "\n"

interfaceAtk :: [Inimigo] -> String
interfaceAtk [] = "\n\n"
interfaceAtk (i:is) = ( (inimigoNome i) ++ "                    " ++              -- 20 espaços
                        (show (inimigoVidaAtual i) ) ++ "/" ++ (show (inimigoVidaMax i) ) ++
                        "\n" ++ (interfaceAtk is) )

escolhaAtaque :: Personagem -> [Inimigo] -> IO Int
escolhaAtaque p list = do
    digite
    system "clear"

    -- CRIA A INTERFACE DA BATALHA
    putStrLn "              #   BATALHA    #\n"

    printaVida p
    putStr (interfaceAtk list)

    putStrLn "\nComo você quer atacar?\n"
    putStrLn "[1] -> Ataque Forte"
    putStrLn "[2] -> Ataque Fraco"
    putStr "\nDigite sua opção: "
    a <- readLn :: IO Int

    if (a == 1) then do
        putStrLn "Você selecionou ataque FORTE.\n"
        return a
    else if (a == 2) then do
        putStrLn "Você selecionou ataque FRACO.\n"
        return a
    else do
        escolhaAtaque p list

auxAtkInim :: Int -> String
auxAtkInim 1 = "\n"
auxAtkInim i = ""

printAtkInimigo :: [Inimigo] -> Int -> String
printAtkInimigo [] i = ""
printAtkInimigo (gp:gpt) i = (auxAtkInim i) ++ ("Atacar: [" ++ (show i) ++ "] "
                            ++ (inimigoNome gp) ++ " " ++
                            (show (inimigoVidaAtual gp)) ++ "/" ++
                            (show (inimigoVidaMax gp)) ++ "\n" ++
                            (printAtkInimigo gpt (i+1)) )

escolheInimigo :: GrupoDeInimigos -> IO Int
escolheInimigo gp = do
    putStrLn (printAtkInimigo (grupoInimigos gp) 1)
    putStr "\nDigite quem você quer atacar: "
    a <- readLn :: IO Int

    if ( (a <= (grupoQuantidade gp)) && (a > 0) ) then
        return (a - 1)
    else
        escolheInimigo gp

entraBatalha :: [Inimigo] -> IO()
entraBatalha grupo = do
    system "clear"
    putStrLn "Você acaba de entrar em uma batalha!!\n"

    putStrLn "Conheça seus inimigos:\n"
    putStrLn (meetYourEnemies grupo)

meetYourEnemies :: [Inimigo] -> String
meetYourEnemies [] = ""
meetYourEnemies (i:is) = (inimigoNome i) ++ " " ++
                         (show (inimigoVidaAtual i) ) ++ "/" ++
                         (show (inimigoVidaMax i) ) ++ "\n" ++
                         "    +-> " ++ (inimigoDescricao i) ++ "\n\n" ++
                         (meetYourEnemies is)




percorreArma1 :: Int -> [Arma] -> String
percorreArma1 indArma (x:[]) = "[" ++ (show (indArma + 1)) ++ "]" ++ " - " ++armaNome x ++ "\n    +->" ++ armaDescricao x  ++ "\nPreço: " ++ show ((armaPreco x)) ++"    Dano:" ++ show(armaDano x) ++ "    Força: " ++ show(armaForca x) ++ "    Agilidade: " ++ show(armaAgilidade x) ++"\n\n" ++  (percorreArma1 (indArma + 1) [])
percorreArma1 indArma [] = "\n************************\n\n[0] -> Cancelar compra.\n\n************************\n\n\n"
percorreArma1 indArma (x:xs) =  "[" ++ (show (indArma + 1)) ++ "]" ++ " "++armaNome x ++ "\n    +->" ++ armaDescricao x  ++ "\nPreço: " ++ show ((armaPreco x)) ++"    Dano:" ++ show(armaDano x) ++ "    Força: " ++ show(armaForca x) ++ "    Agilidade: " ++ show(armaAgilidade x) ++"\n\n" ++  (percorreArma1 (indArma + 1) xs)

percorreArmadura1 :: Int -> [Armadura] -> String
percorreArmadura1 indArmadura (x:[]) = "[" ++ show ((indArmadura) + 1) ++ "]" ++" - "++ (armaduraNome x) ++ "\n   +->" ++ (armaduraDescricao x) ++ "\n" ++ "Preço: " ++ show ((armaduraPreco x)) ++ "     Dano: " ++ show((armaduraArmadura x)) ++ "     Força: " ++ show((armaduraForca x)) ++"     Agilidade: " ++show( (armaduraAgilidade x)) ++"     Vida: " ++show( (armaduraVida x)) ++ "\n\n" ++ (percorreArmadura1 (indArmadura+1) []) 
percorreArmadura1 indArmadura [] = "\n************************\n\n[0] -> Cancelar compra.\n\n************************\n\n\n"
percorreArmadura1 indArmadura (x:xs) = "[" ++ show ((indArmadura) + 1) ++ "]" ++" - "++(armaduraNome x) ++ "\n    +->" ++ (armaduraDescricao x) ++ "\n" ++ "Preço: " ++ show ((armaduraPreco x)) ++"     Dano: " ++ show((armaduraArmadura x)) ++"     Força: " ++ show((armaduraForca x)) ++"     Agilidade: " ++show( (armaduraAgilidade x)) ++"     Vida: " ++show( (armaduraVida x)) ++ "\n\n" ++ (percorreArmadura1 (indArmadura + 1) xs)

percorrePocao1 :: Int -> [Pocao] -> String
percorrePocao1 indPocao (x:[]) = "[" ++ (show (indPocao + 1)) ++ "]" ++" - "++ (pocaoNome x) ++ "\n   +->" ++ (pocaoDescricao x) ++ "\n" ++ "Preço: " ++ show ((pocaoPreco x)) ++ "     Vida: " ++ show((pocaoVida x)) ++ "\n\n"++ (percorrePocao1 (indPocao + 1) [])
percorrePocao1 indPocao [] = "\n************************\n\n[0] -> Cancelar compra.\n\n************************\n\n\n"
percorrePocao1 indPocao (x:xs) = "[" ++ (show (indPocao + 1)) ++ "]" ++" - "++(pocaoNome x) ++ "\n    +->" ++ (pocaoDescricao x) ++ "\n" ++ "Preço: " ++ show ((pocaoPreco x)) ++ "     Vida: " ++ show((pocaoVida x)) ++ "\n\n" ++ (percorrePocao1(indPocao+1) xs)

-- Funções auxiliares para tamanho do array (importadas de Bau)

tamanhouPocao :: [Pocao] -> Int
tamanhouPocao [] = 0
tamanhouPocao (x:[]) = 1
tamanhouPocao (x:xs) = 1 + tamanhouPocao (xs)


tamanhouArmadura :: [Armadura] -> Int
tamanhouArmadura [] = 0
tamanhouArmadura (x:[]) = 1
tamanhouArmadura (x:xs) = 1 + tamanhouArmadura (xs)


tamanhouArma :: [Arma] -> Int
tamanhouArma [] = 0
tamanhouArma (x:[]) = 1
tamanhouArma (x:xs) = 1 + tamanhouArma (xs)

-- fim 

listaArma :: [Arma] -> Personagem -> IO Int
listaArma ar per = do
    system "clear"
    putStr "Seu money -> "
    print (personagemDinheiro per)
    putStrLn ""
    putStr (percorreArma1 0 ar)
    putStr "Digite sua escolha: "
    
    escolhaopcao <- readLn :: IO Int
    let esc = escolhaopcao - 1

    let poiva = armaPreco ( ar !! (esc) )
    let tam = tamanhouArma ar

    if (escolhaopcao > 0 && escolhaopcao <= tam) then do
         if personagemDinheiro per >= poiva then do
            let nome = armaNome ( ar !! (esc) )
            
            putStr "\n"
            estrelinhas
            putStr "\n Você comprou, "
            putStr nome 
            putStrLn " vá até o seu baú \n"
            estrelinhas
            putStr "\n"
            
            digite
            return esc
         else do
            let escolhaopcao = -1
            putStrLn "\n\nVc tá Liso $$$"
            return esc
    else if (escolhaopcao == 0) then
        return (-1)
    else do
        putStrLn "\nDigite uma opção válida."
        listaArma ar per

listaArmadura :: [Armadura] -> Personagem -> IO Int
listaArmadura armadur per = do
    system "clear"
    putStr "Seu money -> "
    print (personagemDinheiro per)
    putStrLn ""
    putStr (percorreArmadura1 0 armadur)
    putStr "Digite sua escolha: "
    
    escolhaopcao <- readLn :: IO Int
    let esc = escolhaopcao - 1

    let poiva = armaduraPreco (armadur !! esc )
    let tam = tamanhouArmadura (armadur)

    if (escolhaopcao > 0 && escolhaopcao <= tam) then do
         if personagemDinheiro per >= poiva then do
            let nome = armaduraNome ( armadur !! esc )
            
            putStr "\n"
            estrelinhas
            putStr "\n Você comprou, "
            putStr nome 
            putStrLn " vá até o seu baú \n"
            estrelinhas
            putStr "\n"
            
            digite
            return esc
         else do
            let escolhaopcao = -1
            putStrLn "\n\nVc tá Liso $$$"
            return esc
    else if (escolhaopcao == 0) then
        return (-1)
    else do
        putStrLn "\nDigite uma opção válida"
        listaArmadura armadur per

listaPocao :: [Pocao] -> Personagem -> IO Int
listaPocao po per = do
    system "clear"
    putStr "Seu money -> "
    print (personagemDinheiro per)
    putStrLn ""
    putStr (percorrePocao1 0 po)
    putStr "Digite sua escolha: "
    
    escolhaopcao <- readLn :: IO Int
    let esc = escolhaopcao - 1

    let poiva = pocaoPreco (po !! esc )
    let tam = tamanhouPocao (po)

    if (escolhaopcao > 0 && escolhaopcao <= tam) then do
        if personagemDinheiro per >= poiva then do
            let nome = pocaoNome ( po !! esc )
            
            putStr "\n"
            estrelinhas
            putStr "\n Você comprou, "
            putStr nome 
            putStrLn " vá até o seu baú \n"
            estrelinhas
            putStr "\n"
            
            digite
            return esc
         else do
            let escolhaopcao = -1
            putStrLn "\n\nVc tá Liso $$$"
            return esc
    else if (escolhaopcao == 0) then
        return (-1)
    else do
        putStrLn "\nDigite uma opção válida."
        listaPocao po per


opcoesBau :: IO Int
opcoesBau = do
    system "clear"
    putStrLn "              #   BAU    #\n"
    putStrLn "[1] -> Troque seu equipamento"
    putStrLn "[2] -> Excluir item da Bau"
    putStrLn "[3] -> Visualizar Equipamento"
    putStrLn "[4] -> Vasculhar sua Bolsa"
    putStrLn "[5] -> Voltar ao menu inicial"
    putStr "\nO que voce quer fazer?... "
    escolha <- readLn :: IO Int

    return escolha

opcoesTroca :: IO Int
opcoesTroca = do
    system "clear"
    putStrLn "              #   TROQUE SEU EQUIPAMENTO    #\n"
    putStrLn "[1] -> Troque sua armadura"
    putStrLn "[2] -> Troque sua arma"
    putStrLn "[3] -> voltar"
    putStr "\nO que voce quer fazer?... "
    escolha <- readLn :: IO Int
    return escolha

opcoesRemove :: IO Int
opcoesRemove = do
    system "clear"
    putStrLn "              #   REMOVA SEU EQUIPAMENTO    #\n"
    putStrLn "[1] -> remova sua armadura"
    putStrLn "[2] -> remova sua arma"
    putStrLn "[3] -> voltar"
    putStr "\nO que voce quer fazer?... "
    escolha <- readLn :: IO Int
    return escolha

opcoesRemoveBolsa :: IO Int
opcoesRemoveBolsa = do
    system "clear"
    putStrLn "              #   BOLSA    #\n\n"
    putStrLn "[1] -> Visualizar poções"
    putStrLn "[2] -> Usar uma poção"
    putStrLn "[3] -> Jogar uma poção fora"
    putStrLn "[4] -> Voltar a batalhar"
    putStr "\n O que voce quer fazer?..."
    escolha <- readLn :: IO Int
    return escolha

opcoesTrocaItens :: IO Int
opcoesTrocaItens = do
  system "clear"
  putStrLn "              #   TROQUE SEU EQUIPAMENTO    #\n"
  putStrLn "[1] -> troque sua armadura"
  putStrLn "[2] -> troque sua arma"
  putStrLn "[3] -> voltar"
  putStrLn "O que voce quer fazer?... "
  escolha <- readLn :: IO Int
  return escolha

vizualizarEquipamento :: Personagem -> IO()
vizualizarEquipamento per = do
    system "clear"
    putStrLn "++++++++++++++++++++++++++++++++++++++++\n"
    putStrLn "          #CONHEÇA O HEROI#              \n"
    putStrLn $ "Nome do Herói: " ++ (personagemNome per) ++ "     HP: " ++ show (personagemVidaAtual per) ++ "/"++ show(personagemVidaMax per)
    putStrLn $ "\n       Dinheiro: " ++ (show (personagemDinheiro per)) ++ "\n"

    putStrLn "Atributos do jogador: "
    putStrLn $ "    Defesa: " ++ show(personagemDefesa per)
    putStrLn $ "    Dano: " ++ show (personagemDano per)
    putStrLn $ "    Força: " ++ show (personagemForca per)
    putStrLn $ "    Agilidade: " ++ show (personagemAgilidade per)

    let arma = personagemArma per
    let armadura = personagemArmadura per

    putStrLn $ "\nSua Arma: " ++ (armaNome arma)
    putStrLn $ "        +> " ++ (armaDescricao arma)
    putStrLn $ "    dano: " ++ show(armaDano arma)
    putStrLn $ "    força: " ++ show(armaForca arma)
    putStrLn $ "    agilidade: " ++ show(armaAgilidade arma)

    putStrLn $ "\nSua Armadura " ++ (armaduraNome armadura)
    putStrLn $ "        +> " ++ (armaduraDescricao armadura)
    putStrLn $ "    armadura: " ++  show (armaduraArmadura armadura)
    putStrLn $ "    força: " ++     show (armaduraForca armadura)
    putStrLn $ "    agilidade: " ++ show (armaduraAgilidade armadura)
    putStrLn $ "    vida: " ++      show (armaduraVida armadura)
    putStrLn "\n\n++++++++++++++++++++++++++++++++++++++++"
    digite



digite :: IO()
digite = do
    putStr "\nDigite enter para continuar..."
    enter
    hSetEcho stdin True
    putStr "\n"

enter :: IO Char
enter = do
    hSetEcho stdin False
    hSetBuffering stdin NoBuffering
    c <- getChar
    return c

enterAux :: IO()
enterAux = do
    enter
    hSetEcho stdin True
    return ()

final :: Personagem -> IO()
final p = do
    system "clear"
    putStrLn "Tecle enter para prosseguir."
    
    enterAux
    putStr "\n\n"
    putStr "+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n"
    putStr "\n"
    
    putStrLn "..."
    enterAux
    putStrLn ("Parabéns " ++ personagemNome p ++ " você derrotou a terrível LIGHT THEME IDE")
    enterAux
    putStrLn "..."
    enterAux
    putStrLn "Mas infelizmente nada acontece."
    enterAux
    putStrLn "..."
    enterAux
    putStrLn "Você ainda pode jogar mais se quiser..."
    enterAux
    putStrLn "\nOu não, sei lá"
    enterAux
    putStrLn "..."
    enterAux
    putStrLn "eh isso\n"
    putStr "+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n"
    putStr "\n\n"

